broadinstitute/single_cell_portal_core
# Refactor study sync logic into `StudySyncService` (SCP-3121)

#### BACKGROUND
With the separation of the API from the more "traditional" Rails controllers, this created areas of the codebase that were prone to duplication.  To address this, separate service libraries were created that could be referenced from both places so that there is only codepath that is exercised, regardless of how many times it is referenced.  This can be seen in practice with classes such as `ClusterVizService`, `ExpressionVizService`, and `FileParseService`.

#### CHANGES
The `StudySyncService` class now handles all main sync functionality, and is referenced in both `StudiesController` and `Api::V1::StudiesController`.  These controllers now only contain code to set the various variables needed for their associated views, and the actual business logic of synchronizing a study with a Terra workspace only lives in `StudySyncService`.  It also now has much more extensive test coverage than the existing controllers had, since the majority of these new helper methods can be tested directly using mocks/stubbing, rather than a full integration test done at the HTTP request level.  Also, this service class is simpler than the preexisting code as it does not rely on the in-memory data structures that the older code utilized for tracking state.

#### MANUAL TESTING
1. Sign into Terra, and create a new workspace in your default development Terra billing project
2. Click the "Open bucket in browser" link under "Cloud Information"
3. Upload several files from `test/test_data` (it doesn't matter which files, just that you upload more than one)
4. Pull this branch, boot as normal and then sign in
5. Click "Create a study" in the top menu bar, and then select "Yes" for **Use an existing workspace**, and put in the name of the workspace for **Existing Terra workspace** (make sure you select the correct Terra billing project as well)
6. Click "Save study", and validate that the sync page loads eventually
7. Validate that the files you uploaded show up in the page as unsynced files

_Note: it is not required that you save the files or parse them as none of that logic has been altered as a part of this PR._
